### PR TITLE
Initial commit of chalice test client

### PR DESCRIPTION
**Pull request by jamesls**
Allow you to test all the chalice decorators
we support with the exception of websockets.

Proposal: #1468.

### CODE REVIEW

File: docs/source/api.rst [hunk 1]
```diff
@@ -1237,3 +1237,167 @@ Websockets
...
+
+      Invoke a Lambda function by name.  The name should match
+      the resource name of the function.  This is typically the
+      name of the python function unless an explicit ``name=``
+      kwarg is provided when registering the function.
+
+      Returns an :class:`InvokeResponse` instance.
+
+
+.. class:: TestHTTPCLient(import_name)
```
Comment (kyleknap) [inline docs/source/api.rst:1306]: `TestHTTPCLient` -> `TestHTTPClient`

File: chalice/test.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+            'eventSourceARN': 'arn:aws:sqs:us-west-2:12345:%s' % queue_name,
+            'md5OfBody': '754ac2f7a12df38320e0c5eafd060145',
+            'messageAttributes': {},
+            'messageId': 'message-id',
+            'receiptHandle': 'receipt-handle'
+        } for body in message_bodies]
+        sqs_event = {'Records': records}
+        return sqs_event
+
+    def generate_cw_event(self, source, detail_type,
```
Comment (kyleknap) [inline chalice/test.py:285]: For scheduled events, are you supposed to use this method to get a sample event? I wonder if it makes sense to have a separate method for it.

File: docs/source/topics/testing.rst [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+   @app.lambda_function()
+   def foo(event, context):
+       return {'hello': 'world'}
+
+   @app.lambda_function()
+   def bar(event, context):
+       return {'event': event}
+
+
+Here's how you can test these functions with the test client:
```
Comment (kyleknap) [inline docs/source/topics/testing.rst:32]: What is the file path for these tests? Is it under a `test_app.py` file? I'm wondering if this guide should help suggest where to write tests as well. I also wonder if we help give direction on how to write tests in a separate directory from the `app.py` file in the case they want a `tests` directory.

File: chalice/test.py (hunk 2 above)
Comment (jamesls) [reply chalice/test.py:285]: Yeah the intent is to use this method.  It's the same thing as a cloudwatch event without any additional detail (https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/EventTypes.html#schedule_event_type):

```
{
  "id": "53dc4d37-cffa-4f76-80c9-8b7d4a4d2eaa",
  "detail-type": "Scheduled Event",
  "source": "aws.events",
  "account": "123456789012",
  "time": "2019-10-08T16:53:06Z",
  "region": "us-east-1",
  "resources": [ "arn:aws:events:us-east-1:123456789012:rule/MyScheduledRule" ],
  "detail": {}
}
```

File: docs/source/topics/testing.rst (hunk 3 above)
Comment (jamesls) [reply docs/source/topics/testing.rst:32]: I can update this.  It doesn't really matter where you put them, either a top level `test_app.py` or a `tests/test_app.py`.  I'll update the docs to recommend a `tests/test_app.py` along with instructions on how you might run these tests with pytest.

**Comment by jamesls** [comment]
@kyleknap Feedback incorporated.

**Pull request merged**